chapters/high_level_shader_language_comparison.adoc

```diff
@@ -0,0 +1,972 @@
+|====
+| *Semantic* | *Description* | *Type*
+| BINORMAL[n] | Binormal | float4
+| BLENDINDICES[n] | Blend indices | uint
+| BLENDWEIGHT[n] | Blend weights | float
+| COLOR[n] | Diffuse and specular color | float4
+| NORMAL[n] | Normal vector | float4
+| POSITION[n] | Vertex position in object space. | float4
+| POSITIONT	Transformed vertex position | float4
+| PSIZE[n] | Point size | float
+| TANGENT[n] | Tangent | float4
+| TEXCOORD[n] | Texture coordinates | float4
+|====
+
+`+n+` is an optional integer between 0 and the number of resources supported.
+
+link:https://learn.microsoft.com/en-us/windows/win32/direct3dhlsl/dx-graphics-hlsl-semantics[source]
+
+=== Shader outputs
+
+==== Passing data between stages
+
+E.g. for vertex and tessellations shaders.
+
+===== GLSL
+
+[source,glsl]
+----
+layout (location = <location-index>) out/in <type> <name>;
+----
+
+Example:
+[source,glsl]
+----
+layout (location = 0) out vec3 outNormal;
+layout (location = 1) out vec3 outColor;
+layout (location = 2) out vec2 outUV;
+layout (location = 3) out vec3 outViewVec;
+
+void main() {
+    gl_Position = vec4(inPos, 1.0);
+    outNormal = inNormal;
+}
+----
+
+===== HLSL
+
+[source,hlsl]
+----
+[[vk::location(<location-index>)]] <type> <name> : <semantic-type>;
+----
+
+Example:
+[source,hlsl]
+----
+struct VSOutput
+{
+	                float4 Pos : SV_POSITION;
+[[vk::location(0)]] float3 Normal : NORMAL;
+[[vk::location(1)]] float3 Color : COLOR;
+[[vk::location(2)]] float2 UV : TEXCOORD0;
+[[vk::location(3)]] float3 ViewVec : TEXCOORD1;
+}
+
+VSOutput main(VSInput input) {
+    VSOutput output = (VSOutput)0;
+    output.Pos = float4(input.Pos.xyz, 1.0);
+    output.Normal = input.Normal;
+    return output;
+}
+----
+
+==== Writing attachments
+
+For fragment shaders.
+
+===== GLSL
+
+[source,glsl]
+----
+layout (location = <attachment-index>) out/in <type> <name>;
+----
+
+Example:
+[source,glsl]
+----
+layout (location = 0) out vec4 outPosition;
+layout (location = 1) out vec4 outNormal;
+layout (location = 2) out vec4 outAlbedo;
+
+void main() {
+    outPosition = ...
+    outNormal = ...
+    outAlbedo = ...
+}
+----
+
+===== HLSL
+
+[source,hlsl]
+----
+<type> <name> : SV_TARGET<attachment-index>;
+----
+
+Example:
+[source,hlsl]
+----
+struct FSOutput
+{
+	float4 Position : SV_TARGET0;
+	float4 Normal : SV_TARGET1;
+	float4 Albedo : SV_TARGET2;
+};
+
+FSOutput main(VSOutput input) {
+    output.Position = ...
+    output.Normal = ...
+    output.Albedo = ...
+    return output;
+}
+----
+
+=== Push constants
+
+==== GLSL
+
+[source,glsl]
+----
+layout (push_constant) uniform <structure-type> { <members> } <name>
+----
+
+Example:
+[source,glsl]
+----
+layout (push_constant) uniform PushConsts {
+	mat4 matrix;
+} pushConsts;
+----
+
+==== HLSL
+
+[source,hlsl]
+----
+[[vk::push_constant]] <structure-type> <name>;
+----
+
+[source,hlsl]
+----
+struct PushConsts {
+    float4x4 matrix;
+};
+[[vk::push_constant]] PushConsts pushConsts;
```

Comment:
In D3D, push constants must be handled through root signature, otherwise it is a plain cbuffer.